File: godtools-model/src/main/java/org/cru/godtools/model/Followup.java
package org.cru.godtools.model;

import android.support.annotation.Nullable;
import android.text.TextUtils;

import org.ccci.gto.android.common.jsonapi.annotation.JsonApiAttribute;
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiIgnore;
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiType;
import org.keynote.godtools.android.model.Language;

import java.util.Date;
import java.util.Locale;

import static org.cru.godtools.model.Followup.JSON_API_TYPE;

@JsonApiType(JSON_API_TYPE)
public final class Followup extends Base {
    static final String JSON_API_TYPE = "follow_up";
    private static final String JSON_NAME = "name";
    private static final String JSON_EMAIL = "email";
    private static final String JSON_LANGUAGE = "language_id";
    private static final String JSON_DESTINATION = "destination_id";

    @Nullable
    @JsonApiAttribute(name = JSON_NAME)
    private String mName;
    @Nullable
    @JsonApiAttribute(name = JSON_EMAIL)
    private String mEmail;
    @Nullable
    @JsonApiIgnore
    private Locale mLanguageCode;
    @Nullable
    @JsonApiAttribute(name = JSON_LANGUAGE)
    private Long mLanguageId;
    @Nullable
    @JsonApiAttribute(name = JSON_DESTINATION)
    private Long mDestination;
    @Nullable
    @JsonApiIgnore
    private Date mCreateTime = new Date();

    @Nullable
    public String getName() {
        return mName;
    }

    public void setName(@Nullable final String name) {
        mName = name;
    }

    @Nullable
    public String getEmail() {
        return mEmail;
    }

    public void setEmail(@Nullable final String email) {
        mEmail = email;
    }

    @Nullable
    public Locale getLanguageCode() {
        return mLanguageCode;
    }

    public void setLanguageCode(@Nullable final Locale code) {
        mLanguageCode = code;
    }

    public void setLanguage(@Nullable final Language language) {
        mLanguageId = language != null ? language.getId() : null;
    }

    @Nullable
    public Long getDestination() {
        return mDestination;
    }

    public void setDestination(@Nullable final Long destination) {
        mDestination = destination;
    }

    @Nullable
    public Date getCreateTime() {
        return mCreateTime;
    }

    public void setCreateTime(@Nullable final Date time) {
        mCreateTime = time != null ? time : new Date();
    }

    public boolean isValid() {
        return mDestination != null && mLanguageCode != null && !TextUtils.isEmpty(mEmail);
    }
}

File: library/model/src/main/java/org/cru/godtools/model/Followup.kt
package org.cru.godtools.model

import org.ccci.gto.android.common.jsonapi.annotation.JsonApiAttribute
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiIgnore
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiType
import java.util.Date
import java.util.Locale

private const val JSON_API_TYPE_FOLLOWUP = "follow_up"

private const val JSON_NAME = "name"
private const val JSON_EMAIL = "email"
private const val JSON_LANGUAGE = "language_id"
private const val JSON_DESTINATION = "destination_id"

@JsonApiType(JSON_API_TYPE_FOLLOWUP)
class Followup : Base() {
    @JsonApiAttribute(JSON_DESTINATION)
    var destination: Long? = null

    @JsonApiAttribute(JSON_NAME)
    var name: String? = null
    @JsonApiAttribute(JSON_EMAIL)
    var email: String? = null

    @JsonApiIgnore
    var languageCode: Locale? = null
    @JsonApiAttribute(JSON_LANGUAGE)
    private var languageId: Long? = null

    fun setLanguage(language: Language?) {
        languageId = language?.id
    }

    @JsonApiIgnore
    var createTime: Date? = Date()
        set(time) {
            field = time ?: Date()
        }

    val isValid get() = destination != null && languageCode != null && !email.isNullOrEmpty()
}
